public/app/features/auth-config/components/ServerDiscoveryModal.tsx

```diff
@@ -0,0 +1,66 @@
+import { useForm } from 'react-hook-form';
+
+import { Button, Input, Field, Modal } from '@grafana/ui';
+
+import { Trans } from '../../../core/internationalization';
+import { ServerDiscoveryFormData } from '../types';
+import { isUrlValid } from '../utils/url';
+
+interface Props {
+  isOpen: boolean | undefined;
+  onClose: () => void;
+  onSuccess: (data: ServerDiscoveryFormData) => void;
+}
+
+export const ServerDiscoveryModal = ({ isOpen, onClose, onSuccess }: Props) => {
+  const {
+    handleSubmit,
+    register,
+    formState: { errors },
+  } = useForm({
+    mode: 'onBlur',
+    defaultValues: {
+      url: '',
+    },
+  });
+
+  const validateUrl = (value?: string) => {
+    if (value === '') {
+      return 'Please enter the .well-known/openid-configuration endpoint for your IdP';
+    }
+
+    if (!isUrlValid(value)) {
+      return 'Please enter a valid URL';
+    }
+
+    return true;
+  };
+
+  return (
+    <Modal title="OpenID Connect Discovery URL" onDismiss={onClose} onClickBackdrop={onClose} isOpen={isOpen}>
+      <form
+        onSubmit={(e) => {
+          e.stopPropagation();
+          return handleSubmit(onSuccess)(e);
+        }}
+      >
+        <Field
+          label="The .well-known/openid-configuration endpoint for your IdP"
+          invalid={!!errors.url}
+          error={errors.url?.message}
+          htmlFor="url"
+        >
+          <Input {...register('url', { validate: validateUrl })} width={80} id="url" />
+        </Field>
+        <Modal.ButtonRow>
+          <Button type="submit" variant="primary">
+            <Trans i18nKey={'oauth.form.server-discovery-modal-submit'}>Submit</Trans>
```

Comment:
I think we need to add a loading state in `ServerDiscoveryField`, pass it to this component and change this text to `Saving...` or smth like that when the URL is sent. I've noticed that sending can take a while and there's no feedback to the user that this action is happening until the modal is closed.  Also maybe we can move the `onSuccess` callback to this component to avoid passing it as a prop?

Reply:
I added a loading state in `ServerDiscoveryField` and updated the submit button text and its enabled state based on it.

I didn't move `onSuccess` to the modal component because `onSuccess` needs `setValue` that would have to be passed instead of `onSuccess` anyway. If you feel strong about it I can do it though.